solr/core/src/test/org/apache/solr/response/transform/TestDeeplyNestedChildDocTransformer.java

```diff
@@ -0,0 +1,227 @@
+  }
+
+  @Test
+  public void testChildFilterJSON() throws Exception {
+    indexSampleData(10);
+    String[] tests = new String[] {
+        "/response/docs/[0]/type_s==[donut]",
+        "/response/docs/[0]/toppings/[0]/type_s==[Regular]",
+    };
+
+    assertJQ(req("q", "type_s:donut",
+        "sort", "id asc",
+        "fl", "*,[child hierarchy=true childFilter='toppings/type_s:Regular']"),
+        tests);
+  }
+
+  @Test
+  public void testGrandChildFilterJSON() throws Exception {
+    indexSampleData(10);
+    String[] tests = new String[] {
+        "/response/docs/[0]/type_s==[donut]",
+        "/response/docs/[0]/toppings/[0]/ingredients/[0]/name_s==[cocoa]"
+    };
+
+    try(SolrQueryRequest req = req("q", "type_s:donut", "sort", "id asc",
+        "fl", "*,[child hierarchy=true childFilter='toppings" + PATH_SEP_CHAR + "ingredients" + PATH_SEP_CHAR + "name_s:cocoa']")) {
+      BasicResultContext res = (BasicResultContext) h.queryAndResponse("/select", req).getResponse();
+      Iterator<SolrDocument> docsStreamer = res.getProcessedDocuments();
+      while (docsStreamer.hasNext()) {
+        SolrDocument doc = docsStreamer.next();
+        int currDocId = Integer.parseInt(((StoredField) doc.getFirstValue("id")).stringValue());
+        assertEquals("queried docs are not equal to expected output for id: " + currDocId, grandChildDocTemplate(currDocId), doc.toString());
+      }
+    }
+
+
+
+    assertJQ(req("q", "type_s:donut",
+        "sort", "id asc",
+        "fl", "*,[child hierarchy=true childFilter='toppings" + PATH_SEP_CHAR + "ingredients" + PATH_SEP_CHAR + "name_s:cocoa']"),
+        tests);
+  }
+
+  @Test
+  public void testSingularChildFilterJSON() throws Exception {
+    indexSampleData(10);
+    String[] tests = new String[] {
+        "/response/docs/[0]/type_s==[cake]",
+        "/response/docs/[0]/lonely/test_s==[testing]",
+        "/response/docs/[0]/lonely/lonelyGrandChild/test2_s==[secondTest]"
+    };
+
+    assertJQ(req("q", "type_s:cake",
+        "sort", "id asc",
+        "fl", "*,[child hierarchy=true childFilter='lonely" + PATH_SEP_CHAR + "lonelyGrandChild" + PATH_SEP_CHAR + "test2_s:secondTest']"),
+        tests);
+  }
+
+  private void indexSampleData(int numDocs) throws Exception {
+    for(int i = 0; i < numDocs; ++i) {
+      updateJ(generateDocHierarchy(i), params("update.chain", "nested"));
+    }
+    assertU(commit());
+  }
+
+  private static String id() {
+    return "" + counter.incrementAndGet();
+  }
+
+  private static String grandChildDocTemplate(int id) {
+    int docNum = id / 8; // the index of docs sent to solr in the AddUpdateCommand. e.g. first doc is 0
+    return "SolrDocument{id=stored,indexed,tokenized,omitNorms,indexOptions=DOCS<id:" + id + ">, type_s=[stored,indexed,tokenized,omitNorms,indexOptions=DOCS<type_s:" + types[docNum % types.length] + ">], name_s=[stored,indexed,tokenized,omitNorms,indexOptions=DOCS<name_s:" + names[docNum % names.length] + ">], " +
```

Comment:
Another option, perhaps easiest, is to pre-process to Solr document before calling toString.  This could remove "noise" fields like root.  If an IndexableField is there then replace with stringValue().

Reply:
Would this method go in the TestUtils or the specific test class?